### PR TITLE
feat: implement augur plugin

### LINKED ISSUE #3341: [Bounty] Integrating Augur Turbo, a prediction market, into Mask plugins
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [x] Intermediate
  - [ ] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 2 weeks
- Bounty Size
  - [x] 400 $MASK or $2000

## Part 1: Problem Description

**Background**
Prediction market is fun to play with friends so naturally it should live on social networks. Mask would love to see all sorts of prediction markets thrive on us. Notable ones including Augur, Omen (powered by gnosis), Polymarket, Reality card etc

Augur turbo is the new version of the 6 year old augur on Matic and currently focus on sports betting

**Solution**
Build an Augur Turbo plugin on Mask. The UI could resemble Dhedge in a card style with different tabs serving different purposes

image
_Dhedge UI_

image
_Detail page for a market on Augur Turbo_

The detail page looks a bit messy and less user friendly for a small card ui on mask plugin. I would recommend default tab with title, market description, odds and the buy/sell button. second page could include odds history and chart and perhaps list of participants, third tab dedicated to providing liquidity. You could also change it to whichever is the best for users.

Augur also has the original eth mainnet version so if you have spare time, do it for the mainnet version as well. 

## Part 2: Resources

- [Bounty Guide](https://github.com/DimensionDev/Maskbook/tree/master/docs/bounty-guide.md)
- https://augur.net/developers
- https://help.augur.net/

## Part 3: Contact and Support
@yisiliu for ideas
@Jack-Works for implementations
Mask Network discord server: https://discord.gg/4Xccj7a75Q
Augur discord server: https://discord.com/invite/aNBTq55

Submission Requirements
Please include demo to show the integration is working.

Winner Announcement Date
Winners will be announced at the conclusion of Gitcoin 10 Hackathon

Judging Criteria
Have a working demo.
Clean code & tech design
Clear UI, easy to use and generally good user experience

### PR DESCRIPTION
**Pull request by iRhonin**
closes #3341

- [x] Detect Augur URL
- [x] Show market info
- [x] Buy/Sell

### CODE REVIEW

**Comment by septs** [comment]
plz resolve conflict and rename the pr title

**Comment by septs** [comment]
plz resolve eslint all errors

**Comment by AinuX** [comment]
Tried the build, but can't buy.

Before execute "Buy" operation, I approved "unlock" in plugin card (https://polygonscan.com/tx/0x2d97ef39e1f70f516ce49fec8888f1f3a53c4cefdcc05d351276731e9fbf8b9e).

Error showed up when "buy", seems that the "approve" operation has some problem.
截屏2021-08-10 22 52 48

Please check, thanks! @iRhonin

**Comment by AinuX** [comment]
Tried with latest build, "unlock" is OK now, but still have some problems.

1. Failed to check transaction receipt after doing unlock/buy/sell - occasional problems
截屏2021-08-11 10 34 11

2.  Data accuracy problem
The accuracy in our plugin and augur website is different.
 

https://user-images.githubusercontent.com/2582974/128966703-f1994866-af07-4ccc-b664-97fdd4366b9c.mp4

**Comment by iRhonin** [comment]
Fixed 2 and 3, about 1, how this happens?

> Tried with latest build, "unlock" is OK now, but still have some problems. @iRhonin
> 
> 1. Failed to check transaction receipt after doing unlock/buy/sell - occasional problems
> 
> 截屏2021-08-11 10 34 11
> 
> 1. Data accuracy problem
>    The accuracy in our plugin and augur website is different.
> 
>  iShot2021-08-11.11.39.48.mp4 
> 1. No original tweet info in shared tweet.
> 
> 截屏2021-08-11 11 55 23
> 
> 截屏2021-08-11 11 57 47

**Comment by AinuX** [comment]
For the #1 problem, after submitting tx, then wait for a second, the error will pop up.

https://user-images.githubusercontent.com/2582974/129130652-6567a439-b438-4eec-ae50-2af7aba1d744.mp4

**Comment by AinuX** [comment]
For the second problem, still have problem in data amount.

https://user-images.githubusercontent.com/2582974/129131034-0f3b7b45-204d-45db-86ce-abcfe3b33038.mp4

**Comment by AinuX** [comment]
3. Plugin can't be rendered if post via mask compose
https://twitter.com/AinuxA/status/1426036947525406721
截屏2021-08-13 12 32 08

4. Some link can't be rendered even post via twitter publisher
https://twitter.com/AinuxA/status/1426033270764892162
截屏2021-08-13 12 32 30

@iRhonin

**Comment by iRhonin** [comment]
@AinuX, about 4, what is the alternative way to do it? I guess detecting only based on the pattern on the URL leads to conflicts, deleting non-Augur links as Augur.

**Comment by iRhonin** [comment]
1,2,3,5 Fixed.
For 4, I updated the `base URL` because the Augur team deployed a new release.
Also added `MMA` market support.
@AinuX

**Comment by AinuX** [comment]
> 1,2,3,5 Fixed.
> For 4, I updated the `base URL` because the Augur team deployed a new release.
> Also added `MMA` market support.
> @AinuX

Latest build is failed, please fix it, thanks!

File: packages/maskbook/src/plugins/Augur/Dashboard/index.tsx [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import type { Plugin } from '@masknet/plugin-infra'
+import { base } from '../base'
+
+const dashboard: Plugin.Dashboard.Definition = {
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Augur/Dashboard/index.tsx:4]: it seems like this entry is not used. can you remove it?

File: packages/maskbook/tsconfig.json [hunk 2]
```diff
@@ -5,7 +5,7 @@
     "outDir": "./dist/",
     "tsBuildInfoFile": "./dist/.tsbuildinfo"
   },
-  "include": ["./src/**/*.tsx", "./src/**/*.ts", "./src/**/*.json"],
+  "include": ["./src/**/*.tsx", "./src/**/*.ts", "./src/**/*.json", "src/plugins/Augur/icons.tss"],
```
Comment (Jack-Works) [inline packages/maskbook/tsconfig.json:8]: please revert this change. it seems like an error.

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/maskbook/src/plugins/Augur/Dashboard/index.tsx (hunk 1 above)
Comment (iRhonin) [reply packages/maskbook/src/plugins/Augur/Dashboard/index.tsx]: Nice catch, removed.

File: packages/maskbook/tsconfig.json (hunk 2 above)
Comment (iRhonin) [reply packages/maskbook/tsconfig.json]: Reverted.

**Comment by AinuX** [comment]
Checked

File: packages/maskbook/package.json [hunk 3]
```diff
@@ -75,6 +75,7 @@
     "date-fns": "^2.23.0",
     "elliptic": "^6.5.3",
     "ethereumjs-util": "^7.1.0",
+    "ethers": "^5.4.4",
```
Comment (guanbinrui) [inline packages/maskbook/package.json:78]: It's not being used?

Comment (iRhonin) [reply packages/maskbook/package.json:78]: It is used on `utils/bmath.ts`

Comment (iRhonin) [reply packages/maskbook/package.json:78]: I tried before adding this library but it is not possible, they have different logic and they are not interchangeable easily.
ethers bignumber only works on integers but bignumber.js uses floating-point arithmetic.

File: packages/maskbook/src/plugins/Augur/utils/sportMarket.ts [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+            isInvalid: i === NO_CONTEST_OUTCOME_ID,
+            isWinner: hasWinner && isSameAddress(shareToken, winner) ? true : false,
+            isFinalNumerator: false, // need to translate final numerator payout hash to outcome
+            shareToken,
+        } as Outcome
+    })
+}
+
+const sportsData = {
+    '2': {
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Augur/utils/sportMarket.ts:193]: Please use enum to mark those magic number

```ts
enum SomeType {
    Type2 = 2
}
const sportsData = { [SomeType.Type2]: { ... } }
```

**Review by Jack-Works — CHANGES_REQUESTED** [review]

**Comment by iRhonin** [comment]
Fixed typos and conflicts,
Don't know how to exactly fix `merge_conflict_check`, can you please help me?
Can you please tell me what is needed for this PR to progress?

File: packages/maskbook/src/plugins/Augur/SNSAdaptor/BuyDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+    const { openDialog: openSettingDialog } = useRemoteControlledDialog(PluginTraderMessages.swapSettingsUpdated)
+    //#endregion
+
+    if (!token || !market || (!loadingAmm && !errorAmm && !ammExchange) || !outcome) return null
+    return (
+        <InjectedDialog
+            className={classes.root}
+            open={open}
+            onClose={onDialogClose}
+            title={market.title + ' ' + outcome?.name}
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Augur/SNSAdaptor/BuyDialog.tsx:284]:
```suggestion
            title={market.title + ' ' + outcome?.name ?? ''}
```

File: packages/maskbook/src/plugins/Augur/SNSAdaptor/SellDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+    }, [account, amount, tokenBalance, estimatedResult])
+    //#endregion
+
+    if (!market || (!loadingAmm && !errorAmm && !ammExchange) || !outcome || !cashToken) return null
+    return (
+        <InjectedDialog
+            className={classes.root}
+            open={open}
+            onClose={onDialogClose}
+            title={market.title + ' ' + outcome?.name}
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Augur/SNSAdaptor/SellDialog.tsx:216]:
```suggestion
            title={market.title + ' ' + outcome?.name ?? ''}
```

File: packages/maskbook/src/plugins/Augur/UI/MarketViewDeck.tsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                <Grid item className={classes.meta}>
+                    <Typography variant="body2" color="textSecondary" className={classes.text}>
+                        {market.endDate.toString()}
+                    </Typography>
+                </Grid>
+                {market.hasWinner ? (
+                    <Grid item className={classes.meta}>
+                        <Typography variant="h6" color="textSecondary" className={classes.text}>
+                            {t('plugin_augur_winner')}: {market.outcomes.find((x) => x.isWinner)?.name}{' '}
+                            <svg width="20" height="20" viewBox="0 0 20 20" fill="none">
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Augur/UI/MarketViewDeck.tsx:52]: Move to a single component.

File: packages/maskbook/src/plugins/Augur/apis/fetchMarketInfo.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
+import type { MarketInfo } from '../types'
+
+export async function fetchMarketInfo(address: string, id: string, url: string) {
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Augur/apis/fetchMarketInfo.ts:3]:
```suggestion
export async function fetchMarketInfo(address: string, id: string, url: string): MarketInfo | undefined {
```

File: packages/maskbook/src/plugins/Augur/apis/fetchMarketTrades.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
+import type { Trade } from '../types'
+
+export async function fetchMarketTrades(address: string, id: string, url: string, fromTimestamp: Number = 0) {
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Augur/apis/fetchMarketTrades.ts:3]:
```suggestion
export async function fetchMarketTrades(address: string, id: string, url: string, fromTimestamp: number = 0) {
```

File: packages/maskbook/src/plugins/Augur/hooks/useMarket.ts [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+            return {
+                ...market,
+                marketType: MarketType.Mma,
+                initialOdds,
+                swapFee,
+                collateral,
+                link,
+                ammExchange: { ...ammExchange, address: AMM_FACTORY_ADDRESS },
+            } as Market
+        } else if (marketInfo.cryptoMarket) {
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Augur/hooks/useMarket.ts:109]: Leave comment?

File: packages/maskbook/src/plugins/Augur/constants/sports.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
+import type { Sport } from '../types'
+import Teams from './teams.json'
+import Fighters from './fighters.json'
+
+export default {
+    '2': {
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Augur/constants/sports.ts:6]: `[SportType.type]`

File: packages/maskbook/src/plugins/Augur/apis/fetchMarketInfo.ts (hunk 8 above)
Comment (guanbinrui) [reply packages/maskbook/src/plugins/Augur/apis/fetchMarketInfo.ts:3]: Less to use `as Type`, prefer `function (): Type`. The latter one is safer.
